packages/hub/src/components/GlobalSearch/index.tsx

```diff
@@ -0,0 +1,175 @@
+"use client";
+import { clsx } from "clsx";
+import { useRouter } from "next/navigation";
+import { FC, useCallback, useEffect, useRef } from "react";
+import { fetchQuery, graphql, useRelayEnvironment } from "react-relay";
+import {
+  components,
+  DropdownIndicatorProps,
+  SelectInstance,
+} from "react-select";
+import AsyncSelect from "react-select/async";
+
+import { SearchIcon } from "@quri/ui";
+
+import { GlobalSearchQuery } from "@/__generated__/GlobalSearchQuery.graphql";
+import { SearchResult$key } from "@/__generated__/SearchResult.graphql";
+import { SearchResult } from "./SearchResult";
+import { SearchResultEdge$key } from "@/__generated__/SearchResultEdge.graphql";
+
+export const Query = graphql`
+  query GlobalSearchQuery($text: String!) {
+    result: search(text: $text) {
+      __typename
+      ... on BaseError {
+        message
+      }
+      ... on QuerySearchConnection {
+        edges {
+          cursor
+          ...SearchResultEdge
+          node {
+            id
+            link
+            object {
+              ...SearchResult
+            }
+          }
+        }
+      }
+    }
+  }
+`;
+
+export type SearchOption =
+  | {
+      type: "object";
+      id: string;
+      link: string;
+      edge: SearchResultEdge$key;
+      object: SearchResult$key;
+    }
+  | {
+      type: "error";
+      message: string;
+    };
+
+const DropdownIndicator = (props: DropdownIndicatorProps<SearchOption>) => {
+  return (
+    <components.DropdownIndicator {...props}>
+      <SearchIcon />
+    </components.DropdownIndicator>
+  );
+};
+
+export const GlobalSearch: FC = () => {
+  const environment = useRelayEnvironment();
+  const router = useRouter();
+
+  const loadOptions = async (text: string): Promise<SearchOption[]> => {
+    const result = await fetchQuery<GlobalSearchQuery>(environment, Query, {
+      text,
+    }).toPromise();
+    if (!result) return [];
+    if (result.result.__typename === "BaseError") {
+      return [
+        {
+          type: "error",
+          message: result.result.message,
+        },
+      ];
+    } else if (result.result.__typename !== "QuerySearchConnection") {
+      return [
+        {
+          type: "error",
+          message: "Unknown error",
+        },
+      ];
+    }
+    return result.result.edges.map((edge) => ({
+      type: "object",
+      id: edge.node.id,
+      link: edge.node.link,
+      edge,
+      object: edge.node.object,
+    }));
+  };
+
+  // https://github.com/JedWatson/react-select/discussions/4669#discussioncomment-1994888
+  const ref = useRef<SelectInstance<SearchOption> | null>(null);
+
+  const handleKeyPress = useCallback((event: KeyboardEvent) => {
+    if (event.key === "Escape") {
+      // I assume this is fast because it will be called on each "Escape" press, even when the search is not focused.
+      ref.current?.blur();
+    }
+    if (event.metaKey && event.key === "k") {
```

Comment:
Global `Cmd+k` hotkey!

Reply:
nice!

Reply:
(Minor)
I'd prefer it if there could be some other, more easily searchable file for this section to go in. Seems messy to have keyboard shortcuts just spread throughout our code everywhere.

Reply:
Yep, this could be a custom React hook.

Reply:
I've added `useGlobalShortcut` hook.